auth-context-setup

balzhan-kanatbek opened the pull request.

balzhan-kanatbek committed:
feat: enhance authentication flow in App component

- Added Sign Up and Login pages to the navigation.
- Integrated authentication context to manage user login state.
- Implemented logout functionality to reset the current page to home.
- Updated rendering logic to conditionally display ProfilePage based on authentication status.

balzhan-kanatbek committed:
feat: implement authentication context and user interface for login and signup

balzhan-kanatbek committed:
feat: update authentication and quote management features

- Added Firebase integration for user authentication.
- Refactored AuthContextProvider to handle login, signup, and Google sign-in.
- Introduced MyCollectionPage for users to view their liked and created quotes.
- Updated MainPage and QuoteCard components to conditionally render based on authentication status.
- Removed ProfilePage and adjusted routing logic accordingly.
- Enhanced user experience with password visibility toggle and improved error handling in login/signup forms.
- Updated package.json and .gitignore to include necessary configurations.

balzhan-kanatbek committed:
fix: update text for clarity on MyCollectionPage and SignUpPage

- Modified the description on MyCollectionPage to include "create" in the context of saving and viewing quotes.
- Changed the prompt on SignUpPage from "Sign In" to "Log In" for consistency in terminology.

annasvst commented on src/App.tsx:
```diff
@@ -1,35 +1,156 @@
 import { useState } from "react";
-import { ProfilePage } from "./pages/ProfilePage/index";
+import { MyCollectionPage } from "./pages/MyCollectionPage/index";
```
The point of naming component files index.js/jsx/ts/tsx is that when we import them, we don't need to type 'index', just use the folder name :) 

```suggestion
import { MyCollectionPage } from "./pages/MyCollectionPage";
```

However, since you now use Tailwind for styling, it's better not to create folders for each component. Creating folders is only needed when you need to group multiple files. With current structure, just create a file (e.g. MyCollectionPage.tsx) for each component.

annasvst replied:
You don't need to change anything in the current PR, you can update it in the following assignments.

annasvst commented on src/App.tsx:
```diff
@@ -1,35 +1,156 @@
 }
 
-const allPages = Object.values(Page);
+const publicPages = [Page.home, Page.myCollection];
```
Normally, a page like 'my collection' should not be a public page but should be a page for logged in users.

annasvst commented on src/App.tsx:
```diff
@@ -1,35 +1,156 @@
       {currentPage === Page.home && <MainPage />}
-      {currentPage === Page.profile && <ProfilePage />}
+      {currentPage === Page.myCollection && <MyCollectionPage handlePageChange={handlePageChange} />}
+      {currentPage === "Sign Up" && <SignUpPage onLoginClick={() => setCurrentPage("Login")} onSignUpSuccess={() => handlePageChange(Page.home)} />}
```
Add Sign up and Login to the enum as well

annasvst commented on src/App.tsx:
```diff
@@ -1,35 +1,156 @@
+        {/* Mobile menu / Desktop navigation */}
+        <ul
+          className={`
+            // Mobile styles
```
It's really hard to understand the code in the component due to high amount of styles. Would be better to create a dedicated components for navbar elements - then logic would be in the App component but styles would be primarily hidden in these dedicated UI components.

annasvst commented on src/App.tsx:
```diff
@@ -1,35 +1,156 @@
               </button>
             </li>
           ))}
+          {!isAuthenticated && (
```
Your code would be cleaner if you move login and sign up together in the one if case:
```JS
{!isAuthenticated && (
  <>
    <li>...</li>
    <li>...</li>
  </>
)
}
```

annasvst commented on src/AuthContextProvider.tsx:
```diff
@@ -0,0 +1,124 @@
+    dispatch({ type: 'LOGOUT' });
+  };
+
+  const contextValue = {
```
Try to use the same names - either use login everywhere or signIn everywhere. I'd recommend using login/logout words internally since Firebase uses signOut (you won't be able to use that name or would have to rename imported Firebase function).

annasvst commented on src/AuthContextProvider.tsx:
```diff
@@ -0,0 +1,124 @@
+      const result = await signInWithPopup(auth, provider);
+      const user = result.user;
+      const token = await user.getIdToken();
+      dispatch({ type: 'LOGIN', payload: { user: user.email || user.uid, token, uid: user.uid } });
```
It's important to be precise when structuring data: setting user to either email or uid is not very helpful in real life. Instead, it would be better to have 2 separate properties for these values and not have user property at all.

annasvst commented on src/components/QuoteCard/index.tsx:
```diff
@@ -1,89 +1,63 @@
+        {!onProfilePage && (
+          <button
+            className={`p-2 sm:p-3 rounded-full border border-blue-700 text-blue-700 bg-blue-50 transition flex items-center justify-center
+              ${!isAuthenticated ? 'opacity-50 cursor-not-allowed' : 'hover:bg-blue-100'}
```
It's better to style the button for disabled state (https://tailwindcss.com/docs/hover-focus-and-other-states#required-and-disabled) and separately add disabled attribute like this: `disabled={!isAuthenticated}`

annasvst replied:
I see that you already have disabled attribute set, just update the styles 👍

annasvst commented on src/components/QuoteCard/index.tsx:
```diff
@@ -1,89 +1,63 @@
+            className={`p-2 sm:p-3 rounded-full border border-blue-700 text-blue-700 bg-blue-50 transition flex items-center justify-center
+              ${!isAuthenticated ? 'opacity-50 cursor-not-allowed' : 'hover:bg-blue-100'}
+            `}
+            onClick={() => isAuthenticated && handleLike({ quote, author, likeCount: 0 }, !isLiked)}
```
If you add disabled attribute as suggested above, you can change this line to:
```suggestion
            onClick={() => handleLike({ quote, author, likeCount: 0 }, !isLiked)}
```

annasvst replied:
I see the disabled is added already, you can just accept the suggestion.

annasvst commented on src/components/QuoteCard/index.tsx:
```diff
@@ -1,89 +1,63 @@
-  const [isBouncing, setIsBouncing] = useState(false);
+export const QuoteCard = ({ quote, author, onProfilePage, isUserCreated, handleDeleteCreatedQuote }: QuoteCardProps) => {
+  const likedQuotes = useLikedQuotesContext() ?? [];
+  const { handleLike, handleDislike } = useLikedQuotesDispatchContext() ?? { handleLike: () => {}, handleDislike: () => {} };
```
Something is not right here :) Update the custom hooks you have in the QuotesContextProvider to be as in the AuthContextProvider (useAuth). This way the code will be much cleaner.

annasvst commented on src/components/QuoteCard/index.tsx:
```diff
@@ -1,89 +1,63 @@
+            className={`p-2 sm:p-3 rounded-full border border-blue-700 text-blue-700 bg-blue-50 transition flex items-center justify-center
+              ${!isAuthenticated ? 'opacity-50 cursor-not-allowed' : 'hover:bg-blue-100'}
+            `}
+            onClick={() => isAuthenticated && handleLike({ quote, author, likeCount: 0 }, !isLiked)}
```
According to what I see in the context, handleLike should only take (quote: Quote, shouldLike: boolean) as parameters. Try to start by cleaning up QuotesContextProvider file and then more gradually towards here. 

I'd also recommend to remove the second parameter (shouldLike) from the handleLike function - we can only like once so there's no point in providing shouldLike toggle.

annasvst commented on src/pages/LoginPage/index.tsx:
```diff
@@ -0,0 +1,125 @@
+    setShowPassword(!showPassword);
+  };
+
+  const handleSubmit = async (e: React.FormEvent) => {
```
I cannot see any validation for username and password inputs. Anything that a user may enter is just sent to firebase. It would be better to add some client side validation to guide the user on the username & password length and supported characters, for example.

annasvst commented on src/pages/SignUpPage/index.tsx:
```diff
@@ -0,0 +1,193 @@
+  onSignUpSuccess: () => void;
+}
+
+export const SignUpPage: React.FC<SignUpPageProps> = ({ onLoginClick, onSignUpSuccess }) => {
```
This component is very similar to the LoginPage. Could you combine them or at least re-use some parts?

annasvst commented on src/types.ts:
```diff
@@ -2,4 +2,5 @@ export interface Quote {
   quote: string;
   author: string;
   likeCount: number;
+  userId?: string;
```
Is this userId for the user that created the quote? Better to name it 'createdBy' or 'addedBy'

annasvst commented on src/AuthContextProvider.tsx:
```diff
@@ -0,0 +1,124 @@
+
+interface AuthState {
+  user: string | null;
+  token: string | null;
```
Do you use toke anywhere? If not, you don't need to store it.

If the user is not authenticated, you should not have uid in the AuthState. Therefore isAuthenticated prop is redundant - remove.

balzhan-kanatbek committed:
Resolve merge conflicts

balzhan-kanatbek committed:
feat: add ProfilePage and update navigation logic

- Introduced ProfilePage component for authenticated users.
- Updated App component to include profile navigation and handle page changes using enum values for consistency.
- Adjusted button click handlers for login and signup to utilize enum for better maintainability.

balzhan-kanatbek committed:
feat: refactor App component and implement Navbar for improved navigation

- Replaced inline navigation logic in App component with a dedicated Navbar component for better separation of concerns.
- Updated routing logic to utilize React Router for navigation between pages, including protected routes for authenticated users.
- Refactored authentication handling in AuthContextProvider to streamline user state management.
- Enhanced mobile responsiveness and user experience in the Navbar with toggle functionality.

balzhan-kanatbek committed:
feat: enhance quote management and user authentication features

- Added support for liking and unliking quotes, allowing users to interact with quotes more dynamically.
- Refactored QuotesContextProvider and quotesReducer to manage liked quotes and user interactions effectively.
- Introduced new components for handling authentication forms, improving user experience during login and signup processes.
- Updated MainPage and MyCollectionPage to display liked and created quotes, enhancing user engagement.
- Implemented reusable components for input fields and social sign-in buttons to streamline form handling.
- Added routing updates to ensure seamless navigation between pages, including user authentication checks.
